File: examples/console/src/console.js
import util from "node:util";

function runImpl() {
     console.count();
    console.log("logged", "message", 1, 2, {key: "value"});
    console.trace("This is a trace message");
    console.debug("This is an debug message");
    console.info("This is an info message");
    console.warn("This is a warning message");
    console.error("This is an error message");
    console.count();
    console.assert(1 === 2, "This is an assertion failure");
    console.assert(1 === 1, "This should not be printed");
    console.group("Group 1");
    console.log("Inside Group 1");
    console.groupCollapsed("Group 2");
    console.log("Inside Group 2");
    console.groupEnd();
    console.groupEnd();
    console.count();
    console.count("test");
    console.countReset();
    console.count("test");
    console.count();
    console.dir({key: "value", nested: {a: 1, b: 2}}, {depth: 2, colors: true});
    console.dir({key: "value", nested: {a: 1, b: 2}}, {colors: false});
    console.table(["apples", "oranges", "bananas"]);
    const people = [
        ["Tyrone", "Jones"],
        ["Janet", "Smith"],
        ["Maria", "Cruz"],
    ];
    console.table(people);

    function Person(firstName, lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    const me = new Person("Tyrone", "Jones");
    console.table(me);

    const tyrone = new Person("Tyrone", "Jones");
    const janet = new Person("Janet", "Smith");
    const maria = new Person("Maria", "Cruz");

    console.table([tyrone, janet, maria], ["firstName"]);
    
    // Map examples
    console.log("--- Map Examples ---");
    
    const emptyMap = new Map();
    console.log("Empty map:", emptyMap);
    
    const simpleMap = new Map();
    simpleMap.set("name", "John");
    simpleMap.set("age", 30);
    simpleMap.set("city", "New York");
    console.log("Simple map:", simpleMap);
    
    const mixedMap = new Map();
    mixedMap.set("string", "value");
    mixedMap.set(42, "number key");
    mixedMap.set({ key: "obj" }, "object key");
    console.log("Map with mixed keys:", mixedMap);
    
    const nestedMap = new Map();
    nestedMap.set("user", { name: "Alice", age: 25 });
    nestedMap.set("config", { debug: true, timeout: 5000 });
    console.log("Map with object values:", nestedMap);
    
    const mapWithArray = new Map();
    mapWithArray.set("colors", ["red", "green", "blue"]);
    mapWithArray.set("numbers", [1, 2, 3, 4, 5]);
    console.log("Map with array values:", mapWithArray);
    
    console.log("--- Map in dir() ---");
    console.dir(simpleMap, { colors: false });
    
    console.log("--- Map in inspect() ---");
    console.log(util.inspect(nestedMap, { depth: 2 }));
    
    // Set examples
    console.log("--- Set Examples ---");
    
    const emptySet = new Set();
    console.log("Empty set:", emptySet);
    
    const simpleSet = new Set();
    simpleSet.add("apple");
    simpleSet.add("banana");
    simpleSet.add("cherry");
    console.log("Simple set:", simpleSet);
    
    const mixedSet = new Set();
    mixedSet.add("string");
    mixedSet.add(42);
    mixedSet.add({ key: "obj" });
    mixedSet.add([1, 2, 3]);
    console.log("Set with mixed types:", mixedSet);
    
    const nestedSet = new Set();
    nestedSet.add({ name: "Alice", age: 25 });
    nestedSet.add({ name: "Bob", age: 30 });
    console.log("Set with objects:", nestedSet);
    
    console.log("--- Set in dir() ---");
    console.dir(simpleSet, { colors: false });
    
    console.log("--- Set in inspect() ---");
    console.log(util.inspect(nestedSet, { depth: 2 }));
    
    // WeakSet examples
    console.log("--- WeakSet Examples ---");
    const weakSet = new WeakSet();
    const obj1 = { id: 1 };
    const obj2 = { id: 2 };
    weakSet.add(obj1);
    weakSet.add(obj2);
    console.log("WeakSet:", weakSet);
    
    // WeakMap examples
    console.log("--- WeakMap Examples ---");
    const weakMap = new WeakMap();
    const key1 = { name: "Alice" };
    const key2 = { name: "Bob" };
    weakMap.set(key1, "value1");
    weakMap.set(key2, "value2");
    console.log("WeakMap:", weakMap);
    
    console.time("time-test");
    setTimeout(() => {
        console.timeLog("time-test", "after 1 second");
        setTimeout(() => {
            console.timeLog("time-test", "after 2 seconds");
            console.timeEnd("time-test");
        }, 1000);
    }, 1000);
    }
    
    export const run = runImpl;

File: tests/runtime/console.rs
use crate::common::{CompiledTest, invoke_and_capture_output};
use indoc::formatdoc;
use test_r::{inherit_test_dep, test};

inherit_test_dep!(
    #[tagged_as("console")]
    CompiledTest
);

#[test]
async fn console(#[tagged_as("console")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (r, output) = invoke_and_capture_output(compiled.wasm_path(), None, "run", &[]).await;
    let _ = r?;

    println!("{output}");

    // Removing the last 3 lines which are printed by timers and can have slight differences in the millisecond
    // values

    let lines = output.lines().collect::<Vec<_>>();
    let (output, timer_output) = lines.split_at(lines.len() - 3);
    let output_str = output.join("\n");
    let timer_output = timer_output.join("\n");

    assert_eq!(
        output_str,
        formatdoc!(
            r#"
    default: 1
    logged message 1 2 {{ key: 'value' }}
    TRACE: This is a trace message
    DEBUG: This is an debug message
    INFO: This is an info message
    WARN: This is a warning message
    ERROR: This is an error message
    default: 2
    WARN: Assertion failed: This is an assertion failure
    Group 1
    Inside Group 1
    Group 2
    Inside Group 2
    default: 3
    test: 1
    test: 2
    default: 1
    {colored}
    {{ key: 'value', nested: {{ a: 1, b: 2 }} }}
    ======================
     (index)  │ 0        │
    ----------------------
     0        │ apples   │
     1        │ oranges  │
     2        │ bananas  │
    ======================
    ==============================
     (index)  │ 0       │ 1      │
    ------------------------------
     0        │ Tyrone  │ Jones  │
     1        │ Janet   │ Smith  │
     2        │ Maria   │ Cruz   │
    ==============================
    =======================
     (index)    │ 0       │
    -----------------------
     firstName  │ Tyrone  │
     lastName   │ Jones   │
    =======================
    ========================
     (index)  │ firstName  │
    ------------------------
     0        │ Tyrone     │
     1        │ Janet      │
     2        │ Maria      │
    ========================
    --- Map Examples ---
    Empty map: Map(0) {{}}
    Simple map: Map(3) {{ 'name' => 'John', 'age' => 30, 'city' => 'New York' }}
    Map with mixed keys: Map(3) {{ 'string' => 'value', 42 => 'number key', {{ key: 'obj' }} => 'object key' }}
    Map with object values: Map(2) {{ 'user' => {{ name: 'Alice', age: 25 }}, 'config' => {{ debug: true, timeout: 5000 }} }}
    Map with array values: Map(2) {{ 'colors' => [ 'red', 'green', 'blue' ], 'numbers' => [ 1, 2, 3, 4, 5 ] }}
    --- Map in dir() ---
    Map(3) {{ 'name' => 'John', 'age' => 30, 'city' => 'New York' }}
    --- Map in inspect() ---
    Map(2) {{ 'user' => {{ name: 'Alice', age: 25 }}, 'config' => {{ debug: true, timeout: 5000 }} }}
    --- Set Examples ---
    Empty set: Set(0) {{}}
    Simple set: Set(3) {{ 'apple', 'banana', 'cherry' }}
    Set with mixed types: Set(4) {{ 'string', 42, {{ key: 'obj' }}, [ 1, 2, 3 ] }}
    Set with objects: Set(2) {{ {{ name: 'Alice', age: 25 }}, {{ name: 'Bob', age: 30 }} }}
    --- Set in dir() ---
    Set(3) {{ 'apple', 'banana', 'cherry' }}
    --- Set in inspect() ---
    Set(2) {{ {{ name: 'Alice', age: 25 }}, {{ name: 'Bob', age: 30 }} }}
    --- WeakSet Examples ---
    WeakSet: WeakSet {{ <items unknown> }}
    --- WeakMap Examples ---
    WeakMap: WeakMap {{ <items unknown> }}"#,
            colored = "{ key: \u{1b}[32m'value'\u{1b}[39m, nested: { a: \u{1b}[33m1\u{1b}[39m, b: \u{1b}[33m2\u{1b}[39m } }"
        )
    );

    assert!(timer_output.contains("after 1 second"));
    assert!(timer_output.contains("after 2 seconds"));
    assert!(timer_output.contains("- timer ended"));

    Ok(())
}
